cpp/src/copying/gather.cuh

```diff
@@ -0,0 +1,331 @@
+   * out of bounds
+   * @param stream Optional CUDA stream on which to execute kernels
+   *---------------------------------------------------------------------------**/
+  template <typename column_type, typename iterator_type>
+  void operator()(gdf_column const *source_column,
+		  iterator_type gather_map,
+		  gdf_column *destination_column, bool ignore_out_of_bounds,
+		  bool sync_nvstring_category, cudaStream_t stream) {
+    column_type const *source_data{
+      static_cast<column_type const *>(source_column->data)};
+    column_type *destination_data{
+      static_cast<column_type *>(destination_column->data)};
+
+    gdf_size_type const num_destination_rows{destination_column->size};
+
+    // If gathering in-place or scattering nvstring
+    // (in which case the sync_nvstring_category should be set to true)
+    // allocate temporary buffers to hold intermediate results
+    bool const sync_category =
+      std::is_same<column_type, cudf::nvstring_category>::value &&
+      sync_nvstring_category;
+    bool const in_place = !sync_category && (source_data == destination_data);
+
+    gdf_column temp_src{};
+    gdf_column temp_dest{};
+
+    if (sync_category) {
+      // sync the categories.
+      temp_src = cudf::copy(*source_column);
+      temp_dest = cudf::copy(*destination_column);
+      const gdf_column *const input_columns[2] = {source_column, &temp_dest};
+      gdf_column *output_columns[2] = {&temp_src, destination_column};
+
+      CUDF_EXPECTS(GDF_SUCCESS ==
+		   sync_column_categories(input_columns, output_columns, 2),
+		   "Failed to synchronize NVCategory");
+
+      source_data = static_cast<column_type *>(temp_src.data);
+    }
+
+    rmm::device_vector<column_type> in_place_buffer;
+    if (in_place) {
+      in_place_buffer.resize(num_destination_rows);
+      destination_data = in_place_buffer.data().get();
+    }
+
+    if (ignore_out_of_bounds) {
+      thrust::gather_if(rmm::exec_policy(stream)->on(stream), gather_map,
+			gather_map + num_destination_rows, gather_map,
+			source_data, destination_data,
+			bounds_checker<decltype(*gather_map)>{0, source_column->size});
+    } else {
+      thrust::gather(rmm::exec_policy(stream)->on(stream), gather_map,
+		     gather_map+num_destination_rows, source_data,
+		     destination_data);
+    }
+
+    if (sync_category) {
+      gdf_column_free(&temp_src);
+      gdf_column_free(&temp_dest);
+    }
+
+    // Copy temporary buffers used for in-place gather to destination column
+    if (in_place) {
+      thrust::copy(rmm::exec_policy(stream)->on(stream), destination_data,
+		   destination_data + num_destination_rows,
+		   static_cast<column_type *>(destination_column->data));
+    }
+
+    CHECK_STREAM(stream);
+  }
+};
+
+/**---------------------------------------------------------------------------*
+* @brief Function object for applying a transformation on the gathermap
+* that converts negative indices to positive indices
```

Comment:
Explain what the conversion is, i.e., it wraps negative values back around from the end. Perhaps include an example.

Reply:
It's easier to think of the transformation as undefined for inputs not in the range `[-n, n)`, and mapping `-i` to `-i+n`. Is this OK?

Reply:
Sure, that's a nice way of describing the conversion.